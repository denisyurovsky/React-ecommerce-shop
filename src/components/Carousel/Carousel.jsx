import PropTypes from 'prop-types';
import React from 'react';
import ImageGallery from 'react-image-gallery';

import noImg from '../../assets/images/noImg.png';
import { useWindowSize } from '../../hooks/useWindowSize';

import 'react-image-gallery/styles/scss/image-gallery.scss';
import styles from './Carousel.module.scss';

const HEIGHT_IMG = {
  small: '300px',
  large: '400px',
};

const IMG_HEIGHT = (screenWidth, breakPoint) =>
  screenWidth > breakPoint ? HEIGHT_IMG.large : HEIGHT_IMG.small;
const BREAK_POINT = 600;

const prepareGalleryData = (images, screenWidth, breakPoint) =>
  Array.isArray(images) && images.length
    ? images.map((image) => ({
        original: image,
        originalAlt: image,
        thumbnail: image,
        thumbnailAlt: image,
        originalClass: images.length === 1 ? 'single' : 'set',
        originalHeight: IMG_HEIGHT(screenWidth, breakPoint),
      }))
    : [
        {
          original: noImg,
          originalClass: 'single',
          originalHeight: IMG_HEIGHT(screenWidth, breakPoint),
        },
      ];

const Carousel = ({ images }) => {
  const size = useWindowSize();
  const imageUrls = prepareGalleryData(images, size.width, BREAK_POINT);
  const showBullets = imageUrls.length > 1 && size.width > BREAK_POINT;
  const showThumbnails = imageUrls.length > 1;
  const thumbnailPosition = size.width > BREAK_POINT ? 'left' : 'bottom';
  const showNav = size.width > BREAK_POINT;
  const stylesContainer =
    imageUrls.length > 1
      ? styles.container
      : `${styles.container} ${styles.single}`;

  return (
    <div className={stylesContainer} data-testid={thumbnailPosition}>
      <ImageGallery
        items={imageUrls}
        infinite={false}
        showBullets={showBullets}
        showNav={showNav}
        showPlayButton={false}
        showFullscreenButton={false}
        showThumbnails={showThumbnails}
        thumbnailPosition={thumbnailPosition}
      />
    </div>
  );
};

Carousel.propTypes = {
  images: PropTypes.arrayOf(PropTypes.string).isRequired,
};

export default Carousel;
