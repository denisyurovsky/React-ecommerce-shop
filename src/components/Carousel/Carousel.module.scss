@import "../../assets/styles/colors.scss";
@import "../../assets/styles/breakpoints.scss";

$defaultBreakPont: 768px;

/* stylelint-disable */

.container {
  position: relative;
  margin: 0 auto 2rem;
  width: 100%;
  height: 26rem;
  @media screen and (max-width: $sm) {
    height: 25rem;
    margin: 0 auto;
  }

  :global {

    .image-gallery {
      margin: 0 auto 2rem;
    }

    .image-gallery-slides {
      height: 25rem;
      margin-top: 1rem;
      @media screen and (max-width: $sm) {
        height: 18.75rem;
      }
    }

    .image-gallery-slide.single {
      margin-left: 3.4375rem;

      @media screen and (max-width: $sm) {
        margin-left: 0;
      }
    }

    .image-gallery-bullets {
      bottom: -1.2rem;

      .image-gallery-bullet {
        background-color: $greyDark;
        border: $greyDark;
        width: 0.625rem;
        height: 0.625rem;

        &:hover {
          background-color: $indigo700;
        }

        &:focus {
          background-color: $indigo700;
          border: $indigo700;
        }

        &.active {
          background-color: $indigo700;
          border: $indigo700;
        }
      }
    }

    .image-gallery-thumbnails {
      margin-top: 2rem;
      height: 24rem !important;
      @media screen and (max-width: $sm) {
        margin-top: 0;
      }

      &:hover {
        border: 1px solid $indigo700;
      }

      &:focus {
        border: 1px solid $indigo700;
      }

      &.activ {
        border: 1px solid $indigo700;
      }
    }

    .image-gallery-icon {
      color: $indigo700;
      filter: none;
    }

    .image-gallery-left-nav,
    .image-gallery-right-nav {
      &:disabled {
        color: $greyDark;
      }

      padding: 0;
      top: 0;
      left: -4.15rem;
      right: 0;
      transform: translateY(0) rotateZ(90deg);

      @media screen and (max-width: $defaultBreakPont) {
        left: -3.55rem;
      }      

      .image-gallery-svg {
        height: 2.5rem;
        width: 1.25rem;
      }
    }

    .image-gallery-right-nav {
      top: 98%;
    }
  }
}

.single {
  margin: 0 auto;
  @media screen and (max-width: $sm) {
    height: 19.75rem;
  }
}
/* stylelint-enable */
