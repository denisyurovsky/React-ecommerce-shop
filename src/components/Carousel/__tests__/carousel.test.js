import { render, screen } from '@testing-library/react';
import React from 'react';

import Carousel from '../Carousel';

const emptyArrayOfImages = [];
const singleImage = ['/img_01.jpg'];
const setOfImages = [
  '/img_01.jpg',
  '/img_02.jpg',
  '/img_03.jpg',
  '/img_04.jpg',
];

describe('Carousel component', () => {
  describe('correct render in dependency on quantity of images', () => {
    it('snapshots renders a image with empty array', () => {
      const { asFragment } = render(<Carousel images={emptyArrayOfImages} />);

      expect(asFragment()).toMatchSnapshot();
    });

    it('snapshots renders a single image', () => {
      const { asFragment } = render(<Carousel images={singleImage} />);

      expect(asFragment()).toMatchSnapshot();
    });

    it('snapshots renders a set of images', () => {
      const { asFragment } = render(<Carousel images={setOfImages} />);

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('correct render in dependency on screen width', () => {
    const testWidth = 600;
    const spy = jest.fn();

    beforeAll(() => {
      window.addEventListener('resize', spy);
    });

    it('does not fire resize event by default', () => {
      expect(spy).not.toHaveBeenCalled();
      expect(window.innerWidth).not.toBe(testWidth);
    });

    afterAll(() => {
      window.removeEventListener('resize', spy);
    });

    describe('snapshots', () => {
      it('renders a valid snapshot with default size 600px', () => {
        const { asFragment } = render(<Carousel images={singleImage} />);

        expect(asFragment()).toMatchSnapshot();
      });
    });

    describe('when resize event is fired', () => {
      it('renders correctly under 600px', () => {
        window.innerWidth = 375;
        render(<Carousel images={singleImage} />);
        expect(screen.getByTestId('bottom')).toBeInTheDocument();
      });

      it('renders correctly after 600px', () => {
        window.innerWidth = 800;
        render(<Carousel images={singleImage} />);
        expect(screen.getByTestId('left')).toBeInTheDocument();
      });
    });
  });
});
