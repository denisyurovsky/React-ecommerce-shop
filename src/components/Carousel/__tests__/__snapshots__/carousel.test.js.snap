// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel component correct render in dependency on quantity of images snapshots renders a image with empty array 1`] = `
<DocumentFragment>
  <div
    class="container single"
    data-testid="left"
  >
    <div
      aria-live="polite"
      class="image-gallery"
    >
      <div
        class="image-gallery-content left"
      >
        <div
          class="image-gallery-slide-wrapper left"
        >
          <div
            class="image-gallery-slides"
          >
            <div
              aria-label="Go to Slide 1"
              class="image-gallery-slide  center  single"
              role="button"
              style="display: inherit; transform: translate3d(0%, 0, 0);"
              tabindex="-1"
            >
              <img
                alt=""
                class="image-gallery-image"
                height="400px"
                sizes=""
                src="../../assets/noImg.png"
                srcset=""
                title=""
                width=""
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Carousel component correct render in dependency on quantity of images snapshots renders a set of images 1`] = `
<DocumentFragment>
  <div
    class="container"
    data-testid="left"
  >
    <div
      aria-live="polite"
      class="image-gallery"
    >
      <div
        class="image-gallery-content left"
      >
        <div
          class="image-gallery-thumbnails-wrapper left thumbnails-swipe-vertical"
        >
          <div
            class="image-gallery-thumbnails"
          >
            <div
              aria-label="Thumbnail Navigation"
              class="image-gallery-thumbnails-container"
              style="transform: translate3d(0, 0px, 0); transition: all 450ms ease-out;"
            >
              <button
                aria-label="Go to Slide 1"
                aria-pressed="true"
                class="image-gallery-thumbnail active"
                tabindex="0"
                type="button"
              >
                <span
                  class="image-gallery-thumbnail-inner"
                >
                  <img
                    alt="/img_01.jpg"
                    class="image-gallery-thumbnail-image"
                    src="/img_01.jpg"
                  />
                </span>
              </button>
              <button
                aria-label="Go to Slide 2"
                aria-pressed="false"
                class="image-gallery-thumbnail"
                tabindex="0"
                type="button"
              >
                <span
                  class="image-gallery-thumbnail-inner"
                >
                  <img
                    alt="/img_02.jpg"
                    class="image-gallery-thumbnail-image"
                    src="/img_02.jpg"
                  />
                </span>
              </button>
              <button
                aria-label="Go to Slide 3"
                aria-pressed="false"
                class="image-gallery-thumbnail"
                tabindex="0"
                type="button"
              >
                <span
                  class="image-gallery-thumbnail-inner"
                >
                  <img
                    alt="/img_03.jpg"
                    class="image-gallery-thumbnail-image"
                    src="/img_03.jpg"
                  />
                </span>
              </button>
              <button
                aria-label="Go to Slide 4"
                aria-pressed="false"
                class="image-gallery-thumbnail"
                tabindex="0"
                type="button"
              >
                <span
                  class="image-gallery-thumbnail-inner"
                >
                  <img
                    alt="/img_04.jpg"
                    class="image-gallery-thumbnail-image"
                    src="/img_04.jpg"
                  />
                </span>
              </button>
            </div>
          </div>
        </div>
        <div
          class="image-gallery-slide-wrapper left"
        >
          <button
            aria-label="Previous Slide"
            class="image-gallery-icon image-gallery-left-nav"
            disabled=""
            type="button"
          >
            <svg
              class="image-gallery-svg"
              fill="none"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="1"
              viewBox="6 0 12 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <polyline
                points="15 18 9 12 15 6"
              />
            </svg>
          </button>
          <button
            aria-label="Next Slide"
            class="image-gallery-icon image-gallery-right-nav"
            type="button"
          >
            <svg
              class="image-gallery-svg"
              fill="none"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="1"
              viewBox="6 0 12 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <polyline
                points="9 18 15 12 9 6"
              />
            </svg>
          </button>
          <div
            class="image-gallery-swipe"
          >
            <div
              class="image-gallery-slides"
            >
              <div
                aria-label="Go to Slide 1"
                class="image-gallery-slide  center  set"
                role="button"
                style="display: inherit; transform: translate3d(0%, 0, 0);"
                tabindex="-1"
              >
                <img
                  alt="/img_01.jpg"
                  class="image-gallery-image"
                  height="400px"
                  sizes=""
                  src="/img_01.jpg"
                  srcset=""
                  title=""
                  width=""
                />
              </div>
              <div
                aria-label="Go to Slide 2"
                class="image-gallery-slide  right  set"
                role="button"
                style="display: inherit; transform: translate3d(100%, 0, 0);"
                tabindex="-1"
              >
                <img
                  alt="/img_02.jpg"
                  class="image-gallery-image"
                  height="400px"
                  sizes=""
                  src="/img_02.jpg"
                  srcset=""
                  title=""
                  width=""
                />
              </div>
              <div
                aria-label="Go to Slide 3"
                class="image-gallery-slide   set"
                role="button"
                style="display: inherit; transform: translate3d(200%, 0, 0);"
                tabindex="-1"
              >
                <img
                  alt="/img_03.jpg"
                  class="image-gallery-image"
                  height="400px"
                  sizes=""
                  src="/img_03.jpg"
                  srcset=""
                  title=""
                  width=""
                />
              </div>
              <div
                aria-label="Go to Slide 4"
                class="image-gallery-slide   set"
                role="button"
                style="display: inherit; transform: translate3d(300%, 0, 0);"
                tabindex="-1"
              >
                <img
                  alt="/img_04.jpg"
                  class="image-gallery-image"
                  height="400px"
                  sizes=""
                  src="/img_04.jpg"
                  srcset=""
                  title=""
                  width=""
                />
              </div>
            </div>
          </div>
          <div
            class="image-gallery-bullets"
          >
            <div
              aria-label="Bullet Navigation"
              class="image-gallery-bullets-container"
              role="navigation"
            >
              <button
                aria-label="Go to Slide 1"
                aria-pressed="true"
                class="image-gallery-bullet active"
                type="button"
              />
              <button
                aria-label="Go to Slide 2"
                aria-pressed="false"
                class="image-gallery-bullet"
                type="button"
              />
              <button
                aria-label="Go to Slide 3"
                aria-pressed="false"
                class="image-gallery-bullet"
                type="button"
              />
              <button
                aria-label="Go to Slide 4"
                aria-pressed="false"
                class="image-gallery-bullet"
                type="button"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Carousel component correct render in dependency on quantity of images snapshots renders a single image 1`] = `
<DocumentFragment>
  <div
    class="container single"
    data-testid="left"
  >
    <div
      aria-live="polite"
      class="image-gallery"
    >
      <div
        class="image-gallery-content left"
      >
        <div
          class="image-gallery-slide-wrapper left"
        >
          <div
            class="image-gallery-slides"
          >
            <div
              aria-label="Go to Slide 1"
              class="image-gallery-slide  center  single"
              role="button"
              style="display: inherit; transform: translate3d(0%, 0, 0);"
              tabindex="-1"
            >
              <img
                alt="/img_01.jpg"
                class="image-gallery-image"
                height="400px"
                sizes=""
                src="/img_01.jpg"
                srcset=""
                title=""
                width=""
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Carousel component correct render in dependency on screen width snapshots renders a valid snapshot with default size 600px 1`] = `
<DocumentFragment>
  <div
    class="container single"
    data-testid="left"
  >
    <div
      aria-live="polite"
      class="image-gallery"
    >
      <div
        class="image-gallery-content left"
      >
        <div
          class="image-gallery-slide-wrapper left"
        >
          <div
            class="image-gallery-slides"
          >
            <div
              aria-label="Go to Slide 1"
              class="image-gallery-slide  center  single"
              role="button"
              style="display: inherit; transform: translate3d(0%, 0, 0);"
              tabindex="-1"
            >
              <img
                alt="/img_01.jpg"
                class="image-gallery-image"
                height="400px"
                sizes=""
                src="/img_01.jpg"
                srcset=""
                title=""
                width=""
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
